cirq-core/cirq/transformers/merge_single_qubit_gates.py

```diff
@@ -152,3 +166,191 @@ def merge_func(m1: 'cirq.Moment', m2: 'cirq.Moment') -> Optional['cirq.Moment']:
+        >>> assert new_sweep[0] == cirq.ParamResolver({'a0': -1, 'x0': 1, 'z0': 0, 'cz_exp': 0})
+        >>> assert new_sweep[1] == cirq.ParamResolver({'a0': -0.5, 'x0': 0, 'z0': -1, 'cz_exp': 1})
+
+    Args:
+        circuit: Input circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
+        sweep: Sweep of the symbols in the input circuit, updated Sweep will be returned
+            based on the transformation.
+        atol: Absolute tolerance to angle error. Larger values allow more negligible gates to be
+            dropped, smaller values increase accuracy.
+
+    Returns:
+        Copy of the transformed input circuit.
+    """
+    deep = context.deep if context else False
+
+    # Tag symbolized single-qubit op.
+    symbolized_single_tag = "TMP-TAG-symbolized-single"
+
+    circuit_tagged = transformer_primitives.map_operations(
+        circuit,
+        lambda op, _: (
+            op.with_tags(symbolized_single_tag)
+            if protocols.is_parameterized(op) and len(op.qubits) == 1
+            else op
+        ),
+        deep=deep,
+    )
+
+    # Step 0, isolate single qubit symbols and resolve the circuit on them.
+    single_qubit_gate_symbols: set[sympy.Symbol] = set().union(
+        *[
+            protocols.parameter_symbols(op) if symbolized_single_tag in op.tags else set()
+            for op in circuit_tagged.all_operations()
+        ]
+    )
+    # If all single qubit gates are not parameterized, call the nonparamerized version of
+    # the transformer.
+    if not single_qubit_gate_symbols:
+        return (merge_single_qubit_gates_to_phxz(circuit, context=context, atol=atol), sweep)
+    sweep_of_single: Sweep = _sweep_on_symbols(sweep, single_qubit_gate_symbols)
+    # Get all resolved circuits from all sets of resolvers in sweep_of_single.
+    resolved_circuits = [
+        protocols.resolve_parameters(circuit_tagged, resolver) for resolver in sweep_of_single
+    ]
+
+    # Step 1, merge single qubit gates per resolved circuit, preserving
+    #  the symbolized_single_tag with indexes.
+    merged_circuits: List['cirq.Circuit'] = []
+    for resolved_circuit in resolved_circuits:
+        merged_circuit = index_tags(
+            merge_single_qubit_gates_to_phxz(
+                resolved_circuit,
+                context=context,
+                merge_tags_fn=lambda circuit_op: (
+                    [symbolized_single_tag]
+                    if any(
+                        symbolized_single_tag in set(op.tags)
+                        for op in circuit_op.circuit.all_operations()
+                    )
+                    else []
+                ),
+                atol=atol,
+            ),
+            target_tags={symbolized_single_tag},
+            context=context,
+        )
+        merged_circuits.append(merged_circuit)
+
+    if not all(
```

Comment:
I don't think this should happen here ... this should be a test for the correctness of the transformer

Reply:
It's to validate the input parameters `sweep`, basically all the resolvers in sweep shoulld resolve the circuit to the same structure, otherwise, we can't do the merge here.

If we don't check here, Step 2 of parameterization would crash.

I should probably update the error message as `InvalidArgument("Invalid input `sweep`, different resolvers in sweep resulted in different merged structures.")`